src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs

```diff
@@ -92,9 +102,70 @@ private static void HandleEndObject(ref ReadStack state)
             }
             else
             {
+                // Set directly when handling non-nested preserved array
+                bool setPropertyDirectly = state.Current.IsPreservedArray && !state.Current.IsNestedPreservedArray;
                 state.Pop();
 
-                ApplyObjectToEnumerable(value, ref state);
+                ApplyObjectToEnumerable(value, ref state, setPropertyDirectly);
+            }
+        }
+
+        private static object GetPreservedArrayValue(ref ReadStack state)
+        {
+            // Preserved JSON arrays are wrapped into JsonPreservedReference<T> where T is the original type of the enumerable
+            // and Values is the actual enumerable instance being preserved.
+            JsonPropertyInfo info = GetValuesFromJsonPreservedReference(ref state.Current);
+            object? value = info.GetValueAsObject(state.Current.ReturnValue);
+
+            if (value == null)
+            {
+                ThrowHelper.ThrowJsonException_MetadataPreservedArrayValuesNotFound(info.DeclaredPropertyType);
+            }
+
+            return value!;
+        }
+
+        private static void HandleStartPreservedArray(ref ReadStack state, JsonSerializerOptions options)
+        {
+            // Check we are not parsing into immutable or array.
+            if (state.Current.JsonPropertyInfo!.EnumerableConverter != null)
+            {
+                ThrowHelper.ThrowJsonException_MetadataCannotParsePreservedObjectIntoImmutable(state.Current.JsonPropertyInfo.DeclaredPropertyType);
+            }
+            Type preservedObjType = state.Current.JsonPropertyInfo.GetJsonPreservedReferenceType();
+            if (state.Current.IsProcessingProperty(ClassType.Enumerable))
+            {
+                state.Push();
+                state.Current.Initialize(preservedObjType, options);
+            }
+            else
+            {
+                // Re-Initialize the current frame.
```

Comment:
For array objects, we don't need to `Push` a new frame to the stack, so we just call `Initialize` again since we are going to handle the array at the moment we step into `JsonPreservableArrayReference<T>.Values`.